### PR TITLE
Lisette dev

### PR DESCRIPTION
**Pull request by lisetterz**
Initial commit

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+	docker build -t carsim .
+
+build-nc: ## Build the container without caching
+	docker build --no-cache -t $(APP_NAME) .
+
+run: ## Run container on port configured in `config.env`
+	docker run -i -t --rm --env-file=./config.env -p=$(PORT):$(PORT) --name="$(APP_NAME)" $(APP_NAME)
+
+test:
+	go test wheels/wheels_test.go
```
Comment (Gorillarock) [inline Makefile]: @lisetterz This will not rerun tests which have cached results.

Comment (Gorillarock) [reply Makefile]: also, it will only run the tests in one file, as opposed to all in the project.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+COPY . .
+ 
+# Installs Go dependencies
+RUN go mod download
+ 
+# Builds your app with optional configuration
+RUN go build -o /carsim
+ 
+# Tells Docker which network port your container listens on
+EXPOSE 8080
```
Comment (Gorillarock) [inline Dockerfile]: Since this program does not listen to a port, exposing is unnecessary

File: config.env [hunk 3]
```diff
@@ -0,0 +1,2 @@
+APP_NAME=carsim
+PORT=8080
```
Comment (Gorillarock) [inline config.env]: again, program do not use this port

File: main.go [hunk 4]
```diff
@@ -194,5 +21,9 @@ func main() {
 	door.SetOpen(false)
 	door.SetLocked(true)
 	motor.SetOn(true)
+	motor.SetRPM(2)
+	wheels.SetHalfMotorRPM(motor)
```
Comment (Gorillarock) [inline main.go]: @lisetterz This method is not very extensible.  Perhaps a function which in on the car, and takes a value as input for the c.WheelManipulator.SetRPM function.

File: Dockerfile (hunk 2 above)
Comment (lisetterz) [reply Dockerfile]: Absolutely, I'll get rid of it.

File: Makefile (hunk 1 above)
Comment (lisetterz) [reply Makefile]: Fixed in f604a5eca9a79a1bd5792f3232bb4621f645e3c4

File: Dockerfile (hunk 2 above)
Comment (lisetterz) [reply Dockerfile]: Fixed in fa6979e986a937026be674ca017bb6e422d5320e and f604a5eca9a79a1bd5792f3232bb4621f645e3c4 (removed from config file)

File: config.env (hunk 3 above)
Comment (lisetterz) [reply config.env]: f604a5eca9a79a1bd5792f3232bb4621f645e3c4

File: main.go (hunk 4 above)
Comment (lisetterz) [reply main.go]: I believe you mean something more like this 8224910427e38c24dbaeaf9c11f176001d50cc52 :)

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+	docker build -t carsim .
+
+build-nc: ## Build the container without caching
+	docker build --no-cache -t $(APP_NAME) .
+
+run: ## Run container on port configured in `config.env`
+	docker run -i -t --rm --env-file=./config.env --name="$(APP_NAME)" $(APP_NAME)
+
+test:
+	go test ./...
```
Comment (Gorillarock) [inline Makefile:17]: note: here, a go clean -testcache before go test ./... would help to ensure that cached test results are not used.

File: car/car.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		c.gearbox = &gearbox.Gearbox{
+			Gear: defaultGear,
+		}
+	}
+	return c.gearbox
+}
+
+func (c *Car) SetMotorRPM(motorRpm int) error {
+	if !c.seatbelt.IsEngaged() {
+		return fmt.Errorf("seabelt must be engaged before motor rpm can be increased")
```
Comment (Gorillarock) [inline car/car.go:78]: note, you cannot slow down unless seatbelt is on here.

File: car/car.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	err = c.SetWheels()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Car) ChangeGear(gear int) {
+	c.GearboxManipulator().SetGear(gear)
+	c.SetWheels()
```
Comment (Gorillarock) [inline car/car.go]: Why SetWheels?

File: car/car.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func (c *Car) SetWheels() error {
```
Comment (Gorillarock) [inline car/car.go]: Might be better named if it was UpdateWheels speed or something like that.

File: gearbox/gearbox.go [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+	Gear int
+}
+
+type GearboxManipulator interface {
+	SetGear(int)
+	GetGear() int
+	Print()
+}
+
+func (g *Gearbox) SetGear(gear int) {
```
Comment (Gorillarock) [inline gearbox/gearbox.go:15]: probably want validation here if intending to use the consts from above.

File: gearbox/gearbox.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
+package gearbox
+
+import "fmt"
+
+const (
```
Comment (Gorillarock) [inline gearbox/gearbox.go]: unused

File: main.go [hunk 11]
```diff
@@ -194,5 +22,15 @@ func main() {
...
+	seatbelt.SetEngaged(true)
+	car.SetMotorRPM(4)
 
+	car.Print()
+	fmt.Println("**  Gear change **")
+	car.ChangeGear(4)
+	car.Print()
+
+	fmt.Println("** Accelerate **")
+	car.SetMotorRPM(8)
```
Comment (Gorillarock) [inline main.go:34]: Does not accelerate over time, instead jumps to 8, but this is ok for exercise

File: Makefile (hunk 5 above)
Comment (lisetterz) [reply Makefile:17]: 9312d663bb973c82a731ae40e3f2b1328782ea71

File: car/car.go (hunk 7 above)
Comment (lisetterz) [reply car/car.go]: e9e514e067c97dec9d876e0e3ca9fa2ffad6ef0d

File: car/car.go (hunk 8 above)
Comment (lisetterz) [reply car/car.go]: I'll go for that since it is more clear. e9e514e067c97dec9d876e0e3ca9fa2ffad6ef0d

File: car/car.go (hunk 6 above)
Comment (lisetterz) [reply car/car.go:78]: Right. I just corrected it 194566b3f1a5b781c409c97e4261edf25339b6ea

File: gearbox/gearbox.go (hunk 9 above)
Comment (lisetterz) [reply gearbox/gearbox.go:15]: I finally decided not to use them. d9e86fabc6001384eb73051664cc5a0d5a47cf6f

File: gearbox/gearbox.go (hunk 10 above)
Comment (lisetterz) [reply gearbox/gearbox.go]: d9e86fabc6001384eb73051664cc5a0d5a47cf6f

**Review by Gorillarock — COMMENTED** [review]
Looks good, thank you for your prompt and apt resolutions.  Great job!

**Pull request closed without merging**